File: trunk/src/main/java/net/esper/event/BeanEventAdapter.java
package net.esper.event;

import net.esper.client.ConfigurationEventTypeLegacy;

import java.util.Map;
import java.util.HashMap;

/**
 * A cache and factory class for obtaining {@link EventType} instances and {@link EventBean} instances
 * for Java Bean events. The class caches {@link EventType} instances already known for performance reasons.
 */
public class BeanEventAdapter
{
    private final Map<Class, BeanEventType> typesPerJavaBean;
    private final Map<String, ConfigurationEventTypeLegacy> classToLegacyConfigs;

    /**
     * Ctor.
     * @param classToLegacyConfigs us a map of event type alias to legacy event type config
     */
    public BeanEventAdapter(Map<String, ConfigurationEventTypeLegacy> classToLegacyConfigs)
    {
        typesPerJavaBean = new HashMap<Class, BeanEventType>();

        this.classToLegacyConfigs = new HashMap<String, ConfigurationEventTypeLegacy>();
        if (classToLegacyConfigs != null)
        {
            this.classToLegacyConfigs.putAll(classToLegacyConfigs);
        }
    }

    /**
     * Returns an adapter for the given Java Bean.
     * @param event is the bean to wrap
     * @return EventBean wrapping Java Bean
     */
    public EventBean adapterForBean(Object event)
    {
        Class eventClass = event.getClass();
        EventType eventType = createOrGetBeanType(eventClass);
        return new BeanEventBean(event, eventType);
    }

    /**
     * Creates a new EventType object for a java bean of the specified class if this is the first time
     * the class has been seen. Else uses a cached EventType instance, i.e. client classes do not need to cache.
     * @param clazz is the class of the Java bean.
     * @return EventType implementation for bean class
     */
    public final BeanEventType createOrGetBeanType(Class clazz)
    {
        if (clazz == null)
        {
            throw new IllegalArgumentException("Null value passed as class");
        }

        // Check if its already there
        BeanEventType eventType = typesPerJavaBean.get(clazz);
        if (eventType != null)
        {
            return eventType;
        }

        // Check if we have a legacy type definition for this class
        ConfigurationEventTypeLegacy legacyDef = classToLegacyConfigs.get(clazz.getName());

        eventType = new BeanEventType(clazz, this, legacyDef);
        typesPerJavaBean.put(clazz, eventType);

        return eventType;
    }
}

File: NEsper/events/BeanEventAdapter.cs
using System;
using System.Collections.Generic;
using System.Threading;

using net.esper.client;
using net.esper.compat;
using net.esper.util;

namespace net.esper.events
{
	/// <summary>
	/// A cache and factory class for obtaining 
    /// <seealso cref="EventType"/> instances and <seealso cref="EventBean"/> instances
	/// for object events. The class caches <seealso cref="EventType"/> instances
    /// already known for performance reasons.
	/// </summary>

	public class BeanEventAdapter
	{
		private readonly EDictionary<Type, BeanEventType> typesPerBean;
		private readonly EDictionary<String, ConfigurationEventTypeLegacy> typeToLegacyConfigs;
		private readonly ReaderWriterLock typesPerBeanLock ;

        /// <summary>Default property resolution style.</summary>
        protected PropertyResolutionStyle defaultPropertyResolutionStyle;

		/// <summary> Ctor.</summary>
		
		public BeanEventAdapter()
		{
			typesPerBean = new HashDictionary<Type, BeanEventType>();
			typesPerBeanLock = new ReaderWriterLock();
			typeToLegacyConfigs = new HashDictionary<String, ConfigurationEventTypeLegacy>();
            defaultPropertyResolutionStyle = PropertyResolutionStyle.CASE_SENSITIVE;
		}

        /// <summary>
        /// Initializes a new instance of the <see cref="BeanEventAdapter"/> class.
        /// </summary>
        /// <param name="defaultPropertyResolutionStyle">The default property resolution style.</param>
        public BeanEventAdapter( PropertyResolutionStyle defaultPropertyResolutionStyle )
        {
            this.typesPerBean = new HashDictionary<Type, BeanEventType>();
            this.typesPerBeanLock = new ReaderWriterLock();
            this.typeToLegacyConfigs = new HashDictionary<String, ConfigurationEventTypeLegacy>();
            this.defaultPropertyResolutionStyle = defaultPropertyResolutionStyle;
        }

        /// <summary>
        /// Gets or sets the property resolution style.
        /// </summary>
        /// <value>The property resolution style.</value>

        public virtual PropertyResolutionStyle DefaultPropertyResolutionStyle
        {
            get { return defaultPropertyResolutionStyle; }
            set { defaultPropertyResolutionStyle = value; }
        }

		/// <summary>
		/// Sets the additional mappings for legacy types.
		/// </summary>
		
		public EDictionary<String, ConfigurationEventTypeLegacy> TypeToLegacyConfigs
		{
            get
            {
                throw new NotSupportedException();
            }
			set
			{
				typeToLegacyConfigs.PutAll( value ) ;
			}
		}
		
		/// <summary>
		/// Returns an adapter for the given object.
		/// </summary>
		/// <param name="_event">The event.</param>
		/// <param name="eventId">The optional event id.</param>
		/// <returns>EventBean wrapping object</returns>

        public virtual EventBean AdapterForBean(Object _event, Object eventId)
		{
			Type eventClass = _event.GetType();
			EventType eventType = CreateOrGetBeanType(eventClass);
			return new BeanEventBean(_event, eventType, eventId);
		}

		/// <summary>
		/// Creates a new EventType object for a specified type if this is the first time
		/// the type has been seen. Else uses a cached EventType instance, i.e. client types
        /// do not need to cache.
		/// </summary>
		/// <param name="type">the type of the object.</param>
		/// <returns>EventType implementation for bean class</returns>
		
		public BeanEventType CreateOrGetBeanType(Type type)
		{
		    if (type == null)
		    {
		        throw new ArgumentException("Null value passed as type");
		    }

		    // Check if its already there
		    typesPerBeanLock.AcquireReaderLock(LockConstants.ReaderTimeout);
		    BeanEventType eventType = typesPerBean.Fetch(type);
		    typesPerBeanLock.ReleaseReaderLock();
		    if (eventType != null)
		    {
		        return eventType;
		    }

		    // not created yet, thread-safe create
		    using(new WriterLock(typesPerBeanLock))
		    {
		        eventType = typesPerBean.Fetch(type);
		        if (eventType != null)
		        {
		            return eventType;
		        }

		        // Check if we have a legacy type definition for this class
		        ConfigurationEventTypeLegacy legacyDef = typeToLegacyConfigs.Fetch(type.FullName);

		        string eventTypeId = "TYPE_" + type.FullName;
		        eventType = new BeanEventType(type, this, legacyDef, eventTypeId);
		        typesPerBean[type] = eventType;
		    }

		    return eventType;
		}
	}
}
